### ISSUE TITLE
Media: Error when uploading unknown file type

### ISSUE THREAD

**Issue by spacedmonkey** [GoogleForCreators/web-stories-wp#6310]
## Bug Description

While uploading an example .mov file, I got the following error. 

<img width="574" alt="Screenshot 2021-02-10 at 16 16 08" src="https://user-images.githubusercontent.com/237508/107538346-c7330c80-6bbb-11eb-88f2-9b14af8292f0.png">

File: [Test File](https://user-images.githubusercontent.com/237508/107538396-d619bf00-6bbb-11eb-81cd-7c3b33ed22a1.MOV)

Worth noting this file was recorded on my iPhone 10 and is 1920 × 1080 resolution. 

## Expected Behaviour

Upload should work. 

## Steps to Reproduce

1. Upload drag provided file into editor. See editor. 

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by swissspidy** [comment]
`createObjectURL` is used in `getResourceFromLocalFile` so this is related to #6053 / #6088

**Comment by spacedmonkey** [comment]
> createObjectURL is used in getResourceFromLocalFile so this is related to #6053 / #6088

This only happens for some video files, not all. Other .mov files seem to work fine.

**Comment by spacedmonkey** [comment]
We could do some simple like this. 
```
const getVideoResource = async (file) => {
  const fileName = getFileName(file);
  const mimeType = file.type;
 
  let src = null;
  if( ['video/mp4', 'video/webm'].includes( file.type ) ) {
    const reader = await createFileReader(file);
    src = window.URL.createObjectURL(
        new window.Blob([reader.result], { type: mimeType })
     );
 }
```

or simplier. 

```
const getResourceFromLocalFile = (file) => {
  const type = getTypeFromMime(file.type);
  if (type === 'image') {
    return getImageResource(file);
  }
  if (['video/mp4', 'video/webm'].includes( file.type ) ) {
    return getVideoResource(file);
  }
  return null;
};
```

**Comment by swissspidy** [comment]
Changing the check only work for `video/mp4` and `video/webm` is not a guarantee that `createObjectURL` won't throw.

The only safe way would be to wrap things in a `try { ... } catch { ... }`

It begs the question though what should be returned in case of error. Probably `null`.

But then again, we need to know what we can display instead of the resource (since it could be transcoded). Hence the relation to #6053 / #6088.

**Comment by swissspidy** [comment]
So from what I can see, `createObjectURL` throws on unexpected codecs.

A MOV file with `H.264` codec works just fine, but with a codec like `AAC,HEVC` (which your test fille has) it breaks.

So the mime type check is definitely not enough.

**Comment by spacedmonkey** [comment]
Just wrapping the insides `getVideoResource` doesn't really do anything other than suppress the error of reading the file, so not sure what we should do here. I wonder if there is a library to extract poster image out of a unsupported file /  codex type....

**Comment by spacedmonkey** [comment]
Maybe we could use `ffmpeg` to generate a poster image and replace our existing logic. 
A poster image would give us something to display, a height and width to play with. See this [example](https://trac.ffmpeg.org/wiki/Create%20a%20thumbnail%20image%20every%20X%20seconds%20of%20the%20video).

**Comment by swissspidy** [comment]
> Just wrapping the insides `getVideoResource` doesn't really do anything other than suppress the error of reading the file, so not sure what we should do here.

We need to catch errors and provide some fallback in that case. That's why #6053 / #6088 is so important. It's exactly what these tickets aim to resolve.

> Maybe we could use `ffmpeg` to generate a poster image and replace our existing logic.
> A poster image would give us something to display, a height and width to play with. See this [example](https://trac.ffmpeg.org/wiki/Create%20a%20thumbnail%20image%20every%20X%20seconds%20of%20the%20video).

That's certainly an interesting idea and would allow generating posters for more types of videos. But it does seem a bit unrelated or at least not solve this bug here. Some reservations I'd have about it though:

1. It takes longer than the current method, which is very fast
2. `ffmpeg` could still fail / there could still be errors
3. Once transcoding is done, our current poster generation is very robust and works well.

For video upload, especially (1) and (2) are reasons why we still need some fallback option as mentioned above.

I could see this being useful specifically as progressive enhancement:

1. Drop exotic video file onto library
2. We want to display the video in the media library
2.1. Use the fallback method being evaluated in #6053 / #6088 (e.g. a placeholder icon)
2.2. Try getting local video resource using `createObjectURL`
2.3. Try getting at least a poster image using ffmpeg **`<---- new`**
3. Transcode video using ffmpeg
4. Upload video

(2) and (3) could perhaps happen in parallel. So transcoding can already start while we try to display a nice preview in the library. The downside is a bit of a waste of resources, running 2 ffmpeg worker threads.


----

On a related note:

`getResourceFromLocalFile` tries to create a poster image using `getFirstFrameOfVideo` already, but it's never actually used later and we simply generate it again. Perhaps we could re-use that one?

**Comment by spacedmonkey** [comment]
I think that all three of these issues,  #6053 / #6088 / #6310 are basically the same thing and should / be solved in one PR. 

To be clear, here is my thinking. 
In the `getVideoResource` function have some conditional logic. 

For webm / mp4 files, do as current, generate a blob for src and generate a poster. How this currently works, as been worked on since the start of the editor and has good test coverage. It isn't broken so we should not fix it. 

The problem I am trying to solve here is video files that are not supported by the browser. Only if the video is unplayable by the browser use ffmpeg to generate the poster. Having a poster here has a number of benefits. 
- We can render a video tag with a poster, so the user has something to interact with. 
- We can get a height and width from image, so we can display the correct dimensions / preview in the media library

I believe that if you have a video tag, with a src of null or nothing but a poster, it will still render. So generating a poster meaning we don't `createObjectURL` in this context, solving the original problem I flagged. 

Another option, might be a render a low res, 1-2 second version of the video, so the user has something to interact with. Then in the background, generate the full video and replace a low res preview with the full res video. But I don't think this is much better poster image. 

I think we need to be exploring these images as we need to understand what technically possible here because that we will inform what is the UX should go ahead with here. I don't want to get the state where we have lots of backwards and forwards with UX about placeholders that that are not technically possible.

**Comment by swissspidy** [comment]
> I think that all three of these issues, #6053 / #6088 / #6310 are basically the same thing and should / be solved in one PR.

Yes, that's why I have been pointing at those issues :-)

I think we're pretty much on the same page in regards to the desired outcome (the _what_), just not _how_ we would implement this in detail.

Summarizing this again:

Simply put, if the current method (`createObjectURL`) does not work, we try to generate a poster with `ffmpeg`.

Now, that poster generation is gonna take some time (let's say 30 seconds, just for the sake of it). During that time, there still needs to be _something_ we can display to the user - a placeholder.

I actually sync'd with Sam yesterday about the UX and their suggestions match perfectly with this. I just updated #6088 with that information, based on our discussion here.

**Comment by swissspidy** [comment]
Closing in favor of #6088
